### PR TITLE
mgr/smb: cluster linked  join auth and users/groups resource types

### PR DESCRIPTION
**Pull request by phlogistonjohn**
Depends on #56350 

Drop the embedded versions of the join source and users and groups     configs from the cluster resource type. This means that potentially     sensitive information is *always* stored in a separate resource     from the less-sensitive general cluster config. Future changes can     then make security related choices about those specific resource types     without worrying about having embedded variants inside the cluster resources.
    
The embedded values are moved to 'linked resources'. The JoinAuth and UsersAndGroups resource types gain a new optional field `linked_to_cluster` that binds the resource to a single cluster resource instance, making it exclusive to that cluster and automatically removed if the cluster resource is removed. Making it a suitable replacement for the embedded resources when used from the CLI command `smb cluster create ...`.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/pybind/mgr/smb/handler.py [hunk 1]
```diff
@@ -1173,3 +1173,11 @@ def _cephx_data_entity(cluster_id: str) -> str:
     use for data access.
     """
     return f'client.smb.fs.cluster.{cluster_id}'
+
+
+def rand_name(prefix: str, max_len: int = 18, suffix_len: int = 8) -> str:
+    trunc = prefix[: (max_len - suffix_len)]
```
Comment (adk3798) [inline src/pybind/mgr/smb/handler.py:1176]: it's probably unnecessary, but do we want to validate here that `max_len` is greater than `suffix_len`

File: src/pybind/mgr/smb/enums.py [hunk 2]
```diff
@@ -41,15 +41,12 @@ class AuthMode(_StrEnum):
 
 
 class JoinSourceType(_StrEnum):
-    PASSWORD = 'password'
-    HTTP_URI = 'http_uri'
     RESOURCE = 'resource'
```
Comment (adk3798) [inline src/pybind/mgr/smb/enums.py:44]: is there much point to still having this enum if it only has one option?

**Review by adk3798 — APPROVED** [review]
LGTM

Note: did not review test changes

File: src/pybind/mgr/smb/handler.py (hunk 1 above)
Comment (phlogistonjohn) [reply src/pybind/mgr/smb/handler.py:1176]: Next time I touch that function I'll add that

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-06-10_15:49:42-orch:cephadm-wip-adk-testing-2024-06-10-0759-distro-default-smithi/

failures consisted of

- 18 `in cluster log` type failures, still being cleaned up and a known issue
- 3 failures in the cleanup of rados_api_tests, don't know the tracker number but this is a known issue
- 2 failures in mds_upgrade_sequence, known issue
- 1 failure with staggered upgrade while agent is active, known issue
- lots of dead jobs due to `Error reimaging machines: Failed to power on smithiXXX`, general lab/teuthology issue unrelated to the PRs in the run

Nothing to block merging for most PRs in the run

**Pull request merged**